crates/compression/src/ports.rs

```diff
@@ -0,0 +1,85 @@
+//! Ports this service requires to function.
+
+use fuel_core_types::{
+    fuel_compression::RegistryKey,
+    fuel_tx::{
+        Address,
+        AssetId,
+        CompressedUtxoId,
+        UtxoId,
+        Word,
+    },
+    fuel_types::Nonce,
+};
+
+use crate::tables::RegistryKeyspace;
+
+/// Rolling cache for compression.
+/// Holds the latest state which can be event sourced from the compressed blocks.
+/// The changes done using this trait in a single call to `compress` or `decompress`
+/// must be committed atomically, after which block height must be incremented.
+pub trait TemporalRegistry {
+    /// Reads a value from the registry at its current height.
+    fn read_registry(
+        &self,
+        keyspace: RegistryKeyspace,
+        key: RegistryKey,
+    ) -> anyhow::Result<Vec<u8>>;
+
+    /// Reads a value from the registry at its current height.
+    fn write_registry(
+        &mut self,
+        keyspace: RegistryKeyspace,
+        key: RegistryKey,
+        value: Vec<u8>,
```

Comment:
I think we rely too much on postcard serialization. It is up to the storage to decide what serialization format to use for data.

I think the `TemporalRegistry` should be generic over a `T` type. Something like: 

```rust
pub trait TemporalRegistry<T> {
     /// Reads a value from the registry at its current height.
     fn read_registry(
         &self,
         key: RegistryKey,
     ) -> anyhow::Result<T>;

     /// Reads a value from the registry at its current height.
     fn write_registry(
         &mut self,
         key: RegistryKey,
         value: T,
     ) -> anyhow::Result<()>;

     /// Lookup registry key by the value.
     fn registry_index_lookup(
         &self,
         value: T,
     ) -> anyhow::Result<Option<RegistryKey>>;
 }
```

In this case, the implementation of the `registry_index_lookup` for script and predicate can handle the logic of hashing the value

<img width="876" alt="image" src="https://github.com/user-attachments/assets/6c30cb3e-3073-4fd0-846f-85a596a20a7d">

Reply:
The proposed solution is not viable, since there's no single `T` that covers the input types. It's different for different keyspaces. Currently I'm planning to just have enum over all keyspaces, and that can be used to pass in the values.

Reply:
It is the main idea that there are not single `T`. The register should implement `TemporalRegistry<T>` for each type, that we want to support. As an example: Database can implement `TemporalRegistry<Address>`, `TemporalRegistry<ContractId`, etc.

Reply:
This has some storage overhead, but that's likely not a huge issue.